src/renderer/renderElement.ts

```diff
@@ -114,31 +116,39 @@ const elementWithCanvasCache = new WeakMap<
   ExcalidrawElementWithCanvas
 >();
 
+const shapeCache = new WeakMap<
+  ExcalidrawElement,
+  Drawable | Drawable[] | null
+>();
+
+export function getShapeForElement(element: ExcalidrawElement) {
+  return shapeCache.get(element);
+}
+
+export function invalidateShapeForElement(element: ExcalidrawElement) {
```

Comment:
we could consider removing this function and instead require users to clone the `element` to invalidate the `shape`.

are there ever situations where we'd want to persist the `shape` between `elements`? i could instead use a strong `Map` keyed by the element ID.

Reply:
I haven't reviewed the PR, so I'm not up to speed yet. What do you mean by "clone element to invalidate shape"? We don't want to be cloning elements in tight loops like resizing etc. just to invalidate shape. Or what am I missing?

> are there ever situations where we'd want to persist the shape between elements?

Between *different* elements? Currently, no. I once suggested that duplicating an element (via alt+drag, or copy/paste) should retain the same shape, but the team didn't want to do that (but we may still support it in the future e.g. via copy/paste shape).

Reply:
You answered my question about invalidation. If we didn’t care about allocations we could try to make ExcalidrawElement immutable and get rid of all the potentially error prone calls to invalidate the shape cache. But I see that we want to do this for perf reasons. Are there any environments we support where WeakMap would be too slow?

Also, my other question was if we ever clone an element instance (ie do we ever have two element objects with the same id) and if so should they share the shape cache. Ie before this change if you did newElem = {...oldElem}; the newElem would preserve the shape, after this change it would have to be recomputed.